### PR TITLE
Add child income fields to Annual Income.

### PR DESCRIPTION
**Pull request by joesoftware**

### CODE REVIEW

File: _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx [hunk 1]
```diff
@@ -14,6 +16,26 @@ class AnnualIncomeSection extends React.Component {
   }
 
   // TODO: Figure out best way to enable users to change their response to pension
+
+  updateChildren() {
+    if (this.props.external === undefined || this.props.external.children === undefined) return;
```
Comment (akainic) [inline _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx]: Do you need this check? `this.props.external.children` should never be `undefined`, even if there are no children because it is initialized to an empty array. You're passing the children to this section if the path is `annual-income`, which it would always be for this section. We probably want to write tests to verify that `this.props.external.children` is never undefined instead of checking for it here. If there's ever a case where it could be undefined, that probably means something is wrong with how we've structured the section and `getExternalData`.

File: _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx [hunk 2]
```diff
@@ -14,6 +16,26 @@ class AnnualIncomeSection extends React.Component {
...
+    if (this.props.external.children.length > 0) {
+      for (let i = 0; i < this.props.external.children.length; i++) {
+        const shortName = `${this.props.external.children[i].childFullName.first} ${this.props.external.children[i].childFullName.last}`;
+        if (this.props.data.children[i] === undefined) {
```
Comment (akainic) [inline _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx]: I'm also confused why this is necessary. If `this.props.external.children.length` is greater than 0, when would there be a case where `this.props.data.children[i]` could be `undefined`?

File: _health-care/_js/components/financial-assessment/ChildIncome.jsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+class ChildIncome extends React.Component {
+  render() {
+    const message = 'Please enter only numbers and a decimal point if necessary (no commas or currency signs)';
+
+    return (
+      <div>
+        <h6>{this.props.data.childShortName}</h6>
+        <ErrorableTextInput
+            errorMessage={isBlank(this.props.data.childGrossIncome) || isValidMonetaryValue(this.props.data.childGrossIncome) ? undefined : message}
```
Comment (akainic) [inline _health-care/_js/components/financial-assessment/ChildIncome.jsx]: Good catch that we needed to add `isBlank` to the validation!

File: _health-care/_js/components/form-elements/FixedTable.jsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+
+class FixedTable extends React.Component {
+  render() {
+    const rowElements = this.props.rows.map((obj, index) => {
+      return (
+        <div className="row" key={index}>
+          <div className="small-9 columns">
```
Comment (akainic) [inline _health-care/_js/components/form-elements/FixedTable.jsx]: Let's remove these Foundation classes, they're not necessary here. We can entirely remove line 8 and then just remove the `className="row"` in line 7.

File: _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx (hunk 2 above)
Comment (akainic) [inline _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx]: Nevermind! Now I see what's going on. Was confused by these 2 different arrays of children we're working with.

Comment (akainic) [inline _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx]: I feel like this should not happen in the section level, though. This is basically altering our applicationData state, and that should only happen at the `HealthCareApp` level. I think we need to figure out a way of adding children to the annual income section once children are added in the child information section.

File: _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx [hunk 5]
```diff
@@ -123,26 +147,17 @@ class AnnualIncomeSection extends React.Component {
...
-              label="Children Gross Income"
-              value={this.props.data.childrenGrossIncome}
-              onValueChange={(update) => {this.props.onStateChange('childrenGrossIncome', update);}}/>
 
-          <ErrorableTextInput
-              errorMessage={this.isValidMonetaryValue(this.props.data.childrenNetIncome, message)}
-              label="Children Net Income"
-              value={this.props.data.childrenNetIncome}
-              onValueChange={(update) => {this.props.onStateChange('childrenNetIncome', update);}}/>
+        {this.props.external.children.length > 0 &&
```
Comment (akainic) [inline _health-care/_js/components/financial-assessment/AnnualIncomeSection.jsx]: This should be extracted out above the `render()` function, like we did for `notRequiredMessage`.

**Comment by akainic** [comment]
I see where you're going with this, and it's starting to come together. But I think we need to figure out a better way of building the children array within `annual-income` once children are created in `child-information`. We also need to write tests for `FixedTable`.

File: _health-care/_js/components/financial-assessment/ChildIncome.jsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+
+import ErrorableTextInput from '../form-elements/ErrorableTextInput';
+
+import { isBlank, isValidMonetaryValue } from '../../utils/validations';
+
+// TODO: create unique nodes for each child in applicationData
+
```
Comment (akainic) [inline _health-care/_js/components/financial-assessment/ChildIncome.jsx]: Please add a description of the props like we do in other components.

File: _health-care/_js/components/form-elements/FixedTable.jsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+
```
Comment (akainic) [inline _health-care/_js/components/form-elements/FixedTable.jsx]: Please add a description of the props like we do in other components.

File: _health-care/_js/components/HealthCareApp.jsx [hunk 8]
```diff
@@ -220,16 +219,39 @@ class HealthCareApp extends React.Component {
   getExternalData(applicationData, statePath) {
     switch (statePath[0]) {
       case 'financial-assessment':
+        if (statePath[1] === 'annual-income') {
```
Comment (akainic) [inline _health-care/_js/components/HealthCareApp.jsx]: I think I see what you were saying now about calling functions here instead of just passing down data. A way around this is to call this function within `publishStateChange` instead as a callback on `this.setState`. `this.setState` can take a callback of a function to run once the state has been updated (see line 245 of an example of this). So in `publishStateChange` let's do: 
```javascript
this.setState({ applicationData: newApplicationData }, () = {
  this.updateChildrenIncome();
});
```
And then here we can just pass `childrenIncome: this.state.applicationData['financial-assessment']['annual-income'].children`. I think that'll work.

Comment (joesoftware) [inline _health-care/_js/components/HealthCareApp.jsx]: I'll make a change with the addition of a check of the `propertyPath` to conditionally update.

File: _health-care/_js/components/HealthCareApp.jsx [hunk 9]
```diff
@@ -222,16 +221,39 @@ class HealthCareApp extends React.Component {
       case 'financial-assessment':
         return {
           receivesVaPension: this.state.applicationData['personal-information']['va-information'].receivesVaPension,
-          neverMarried: this.state.applicationData['personal-information']['name-and-general-information'].maritalStatus === '' ||
-            this.state.applicationData['personal-information']['name-and-general-information'].maritalStatus === 'Never Married'
+          neverMarried: this.state.applicationData['personal-information']['name-and-general-information'].maritalStatus === '' || this.state.applicationData['personal-information']['name-and-general-information'].maritalStatus === 'Never Married'
```
Comment (akainic) [inline _health-care/_js/components/HealthCareApp.jsx]: This has now been moved into `_health-care/_js/store/calculated.js`. Take a look at `SpouseInformatinoSection.jsx` to see how it can accessed.

File: _health-care/_js/components/HealthCareApp.jsx [hunk 10]
```diff
@@ -222,16 +221,39 @@ class HealthCareApp extends React.Component {
...
       default:
         return undefined;
     }
   }
 
+  // Make sure the number of children reported have corresponding rows in the annual-income section.
+  // TODO: This should really be looking to match the shortname of instead of just
+  //       just assuming the index between child info and child income match.
+  updateChildrenIncome(children) {
+    const childIncome = this.state.applicationData['financial-assessment']['annual-income'];
```
Comment (akainic) [inline _health-care/_js/components/HealthCareApp.jsx]: `this.state.applicationData` no longer exists, it is now `blankVeteran` in `_health-care/_js/reducers/veteran/index.js`. Since this function updates our store, I think we would want to create an action and reducer for this. The thing I'm not sure about is when this action should be dispatched. Currently we're calling this function whenever a change is made and when the path equals `/financial-assessment/child-information`, so we could match that and dispatch the action when a change is made in `ChildInformationSection` (since we want to add child income rows when a child is added). The function itself will change slightly too, based on how the other reducer functions are structured in creating a new state object and assigning it new values.

File: _health-care/_js/components/HealthCareApp.jsx [hunk 11]
```diff
@@ -222,16 +221,39 @@ class HealthCareApp extends React.Component {
...
+      } else {
+        childIncome.children[i].childShortName = shortName;
+      }
+    }
+    childIncome.children.splice(children.length);
+  }
+
   publishStateChange(propertyPath, update) {
     const newApplicationData = Object.assign({}, this.state.applicationData);
+    if (_.isEqual(propertyPath, ['financial-assessment', 'child-information', 'children'])) {
```
Comment (akainic) [inline _health-care/_js/components/HealthCareApp.jsx]: This will change based on above comment.

File: _health-care/_js/components/HealthCareApp.jsx (hunk 9 above)
Comment (joesoftware) [inline _health-care/_js/components/HealthCareApp.jsx]: Merging master seems to have taken care of this for me.

File: .gitignore [hunk 12]
```diff
@@ -12,3 +12,4 @@ vendor/bundle
 .ruby-gemset
 assets/js/generated/
 .bundle/
+.vscode/
```
Comment (akainic) [inline .gitignore]: What is this from?

File: .eslintrc.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
+module.exports = {
```
Comment (akainic) [inline .eslintrc.js]: I don't think this file should be checked in. In the least it should be added to gitignore, but how did it get generated in the first place?

File: _health-care/_js/components/HealthCareApp.jsx [hunk 14]
```diff
@@ -52,6 +52,7 @@ class HealthCareApp extends React.Component {
     this.context.store.dispatch(ensureFieldsInitialized(path));
     if (validations.isValidSection(path, sectionData)) {
       hashHistory.push(this.getUrl('next'));
+      sectionData.sectionComplete = true;
```
Comment (akainic) [inline _health-care/_js/components/HealthCareApp.jsx]: This is a change from Courtnye's PR #1758, and shouldn't be included in this PR.

File: _health-care/_js/components/Nav.jsx [hunk 15]
```diff
@@ -1,4 +1,5 @@
 import React from 'react';
+import { connect } from 'react-redux';
```
Comment (akainic) [inline _health-care/_js/components/Nav.jsx]: All of these changes are also from #1758 it looks like and shouldn't be included in this PR.

**Comment by akainic** [comment]
Closing this PR because #1852 replaces it.
